File: src/Clones/InspectionProvider.java
package Clones;

import com.intellij.codeInspection.*;
import com.intellij.openapi.project.Project;
import com.intellij.psi.*;
import com.suhininalex.clones.CloneClass;
import com.suhininalex.clones.CloneManager;
import org.jetbrains.annotations.Nls;
import org.jetbrains.annotations.NotNull;

import java.awt.*;
import java.util.List;


public class InspectionProvider extends BaseJavaLocalInspectionTool {

    static AllManager allManager;
    static ProblemsHolder holder;



    synchronized  static public AllManager getViewManager(Project project){
        if (allManager ==null) allManager = new AllManager(project);
        return allManager;
    }

    @Nls
    @NotNull
    @Override
    public String getGroupDisplayName() {
        return "Clones!";
    }

    @NotNull
    public String getShortName() {
        return "CloneDetection";
    }

    @Nls
    @NotNull
    @Override
    public String getDisplayName() {
        return "Suhinin clone detection tool!";
    }


    static volatile boolean visited = false;

    @NotNull
    @Override
    public PsiElementVisitor buildVisitor(@NotNull ProblemsHolder holder, boolean isOnTheFly) {
        if (!visited) {
            synchronized (InspectionProvider.class) {
                if (!visited)
                    EventQueue.invokeLater(() -> getViewManager(holder.getProject()).showProjectClones());
            }
        }
        while (!visited);
        this.holder = holder;
        return new MyVisitor();
    }

    private static class MyVisitor extends JavaElementVisitor {

        @Override
        public void visitMethod(PsiMethod method) {

//            System.out.println("In method: " + Utils.getMethodId(method));
            allManager.cloneManager.updateMethod(method);
            List<CloneClass> clones = getViewManager(method.getProject()).cloneManager.getMethodFilteredClones(method);
            if (!clones.isEmpty())
                holder.registerProblem(method, "Atatata problem!", new MyQuickFix());
        }

    }


    private static class MyQuickFix implements LocalQuickFix {

        @NotNull
        public String getName() {
            return "Clone detected!";
        }

        public void applyFix(@NotNull Project project, @NotNull ProblemDescriptor descriptor) {
            PsiMethod method = (PsiMethod)descriptor.getPsiElement();
            CloneManager cm = getViewManager(project).cloneManager;
            List<CloneClass> clones =  cm.getMethodFilteredClones(method);
            EventQueue.invokeLater(() -> ClonesView.showClonesData(project, clones));
        }

        @NotNull
        public String getFamilyName() {
            return "myfix";
        }
    }
}

File: plugin/src/main/com/suhininalex/clones/ide/InspectionProvider.kt
package com.suhininalex.clones.ide

import com.intellij.codeInspection.*
import com.intellij.openapi.project.Project
import com.intellij.psi.*
import com.suhininalex.clones.core.cloneManager
import com.suhininalex.clones.core.postprocessing.*
import com.suhininalex.clones.core.structures.Clone
import com.suhininalex.clones.core.structures.CloneClass
import java.awt.EventQueue
import com.suhininalex.clones.core.utils.*
import com.suhininalex.clones.ide.toolwindow.CloneViewManager

class InspectionProvider : BaseJavaLocalInspectionTool() {

    override fun getGroupDisplayName() = "Analyze"

    override fun getShortName() = "CloneDetection"

    override fun getDisplayName() = "TreeClone detection"


    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean): PsiElementVisitor =
            CloneInspectionVisitor(holder)

    override fun isInitialized(): Boolean =
            CurrentProject?.cloneManager?.initialized ?: false

}

class CloneInspectionVisitor(val holder: ProblemsHolder) : JavaElementVisitor() {

    override fun visitMethod(method: PsiMethod) {
        val cloneManager = method.project.cloneManager.instance
        val result = cloneManager.getMethodFilteredClones(method)
        result.forEach { cloneClass ->
            cloneClass.clones.filter { it.firstPsi in method  }.forEach { clone ->
                holder.registerProblem(
                        method,
                        "Method may have cloneClasses",
                        ProblemHighlightType.WEAK_WARNING,
                        clone.getTextRangeInMethod(),
                        CloneReport(cloneClass, clone)
                )
            }
        }
    }
}

operator fun PsiElement.contains(element: PsiElement): Boolean =
    containingFile == element.containingFile && element.textRange in textRange

class CloneReport(val cloneClass: CloneClass, val clone: Clone) : LocalQuickFix {

    override fun getName() = "Show cloneClasses for this method ${clone.firstPsi.startLine}:${clone.lastPsi.endLine}"

    override fun getFamilyName() = "CloneReport"

    override fun applyFix(project: Project, descriptor: ProblemDescriptor) {
        EventQueue.invokeLater { CloneViewManager.showClonesData(project, cloneClass) }
    }
}